documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2F<TScalar> UnitX { get; }
+    public static Vector2F<TScalar> UnitY { get; }
+
+    public static Vector2F<TScalar> Zero { get; }
+    public static Vector2F<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2F(TScalar x, TScalar y);
+    public Vector2F(TScalar value);
+    public Vector2F(ReadOnlySpan<TScalar> span);
+
+    public Span<TScalar> AsSpan();
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, TScalar right);
+
+    public static Vector2F<TScalar> operator *(TScalar left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> value);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> value);
+
+    public TScalar LengthSquared { get; }
+    public TScalar Length();
+
+    public Vector2F<TScalar> Dot(Vector2F<TScalar> other);
+    public TScalar DistanceSquared(Vector2F<TScalar> other);
+    public TScalar Distance(Vector2F<TScalar> other);
+    public Vector2F<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
+
+    public static Vector2F<TScalar> Dot (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> DistanceSquared (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Distance (Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> Clamp (Vector2F<TScalar> value, Vector2F<TScalar> min, Vector2F<TScalar> max);
+    public static Vector2F<TScalar> Min (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Max (Vector2F<TScalar> value1, Vector2F<TScalar> value2);
+    public static Vector2F<TScalar> Abs (Vector2F<TScalar> value);
+}
+```
+At the end of this proposal is the definition for the new interfaces used in the these classes.
+
+# Matrix Types
+
+This proposal includes the following matrix types:
+- Matrix2x2F
+- Matrix2x2I
+- Matrix2x3F
+- Matrix2x3I
+- Matrix2x4F
+- Matrix2x4I
```

Comment:
What are `2x3` and `2x4` used for?

Reply:
¯\_(ツ)_/¯ I have no clue, I just was taking from the list I was given,

Reply:
I would expect maybe a specialized `WriteTransposed` or similar in that case, since the primary scenario here is to store to GPU memory as part of a copy, you don't actually need the `2x3` at any point.

Reply:
update: opengl bindings use matrix types 2x2 up to 4x4 with all combinations thereof in parameters and returns so they are all needed from that perspective

Reply:
Can you give an example of such an API?

AFAIK, OpenGL itself doesn't have a built-in matrix type on its native side (it defers everything to GLM, which is separate). GLSL supports a little bit of arbitrary matrix sizes, but noting its `mat#x#` is `CxR`, which is backwards from standard convention. SPIR-V and HLSL then support them as well, but I don't think I've ever seen them used in practice.

Notably, even for float/double, they would basically be "anti-patterns" and users would easily fall into a perf pit of failure by using them, hence why I think it'd be "better" to have the shape that does work well, and to encourage exposing `LoadTransposed` and `StoreTransposed` to allow users to efficiently handling "transpose + copy to/from memory".